# EBAC - Regressão II - regressão múltipla

## Tarefa I

#### Previsão de renda

Vamos trabalhar com a base 'previsao_de_renda.csv', que é a base do seu próximo projeto. Vamos usar os recursos que vimos até aqui nesta base.

|variavel|descrição|
|-|-|
|data_ref                | Data de referência de coleta das variáveis |
|index                   | Código de identificação do cliente|
|sexo                    | Sexo do cliente|
|posse_de_veiculo        | Indica se o cliente possui veículo|
|posse_de_imovel         | Indica se o cliente possui imóvel|
|qtd_filhos              | Quantidade de filhos do cliente|
|tipo_renda              | Tipo de renda do cliente|
|educacao                | Grau de instrução do cliente|
|estado_civil            | Estado civil do cliente|
|tipo_residencia         | Tipo de residência do cliente (própria, alugada etc)|
|idade                   | Idade do cliente|
|tempo_emprego           | Tempo no emprego atual|
|qt_pessoas_residencia   | Quantidade de pessoas que moram na residência|
|renda                   | Renda em reais|

In [43]:
import pandas as pd
import numpy as np
from seaborn import load_dataset
import seaborn as sns
import matplotlib.pyplot as plt

import patsy
import statsmodels.api as sm
import statsmodels.formula.api as smf

In [44]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [45]:
df = pd.read_csv('/content/drive/MyDrive/previsao_de_renda.csv')

In [46]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15000 entries, 0 to 14999
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0             15000 non-null  int64  
 1   data_ref               15000 non-null  object 
 2   id_cliente             15000 non-null  int64  
 3   sexo                   15000 non-null  object 
 4   posse_de_veiculo       15000 non-null  bool   
 5   posse_de_imovel        15000 non-null  bool   
 6   qtd_filhos             15000 non-null  int64  
 7   tipo_renda             15000 non-null  object 
 8   educacao               15000 non-null  object 
 9   estado_civil           15000 non-null  object 
 10  tipo_residencia        15000 non-null  object 
 11  idade                  15000 non-null  int64  
 12  tempo_emprego          12427 non-null  float64
 13  qt_pessoas_residencia  15000 non-null  float64
 14  renda                  15000 non-null  float64
dtypes:

1. Ajuste um modelo para prever log(renda) considerando todas as covariáveis disponíveis.
    - Utilizando os recursos do Patsy, coloque as variáveis qualitativas como *dummies*.
    - Mantenha sempre a categoria mais frequente como casela de referência
    - Avalie os parâmetros e veja se parecem fazer sentido prático.  


2. Remova a variável menos significante e analise:
    - Observe os indicadores que vimos, e avalie se o modelo melhorou ou piorou na sua opinião.
    - Observe os parâmetros e veja se algum se alterou muito.  


3. Siga removendo as variáveis menos significantes, sempre que o *p-value* for menor que 5%. Compare o modelo final com o inicial. Observe os indicadores e conclua se o modelo parece melhor.
    

***1 - Ajuste um modelo para prever log(renda) considerando todas as covariáveis disponíveis***

In [47]:
# excluindo variáveis que, por sua natureza, não devem fazer parte da construção do modelo

df = df.drop(columns=['Unnamed: 0', 'id_cliente'])

In [48]:
# criando matriz de design (e a matriz y) utilizando o Patsy, para um modelo em renda, explicado pelas demais variáveis

y, X = patsy.dmatrices('renda ~ data_ref + sexo + posse_de_veiculo + posse_de_imovel + qtd_filhos + tipo_renda + educacao + estado_civil + tipo_residencia + idade + tempo_emprego + qt_pessoas_residencia', df)

reg = sm.OLS(y,X).fit()

reg.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  renda   R-squared:                       0.257
Model:                            OLS   Adj. R-squared:                  0.255
Method:                 Least Squares   F-statistic:                     112.7
Date:                Sat, 24 Jun 2023   Prob (F-statistic):               0.00
Time:                        21:33:52   Log-Likelihood:            -1.2878e+05
No. Observations:               12427   AIC:                         2.576e+05
Df Residuals:                   12388   BIC:                         2.579e+05
Df Model:                          38                                         
Covariance Type:            nonrobust                                         
====================================================================================================
                                       coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------------------------
Intercept                        -5879.4874   2515.712     -2.337      0.019   -1.08e+04    -948.300
data_ref[T.2015-02-01]             -50.5760    377.087     -0.134      0.893    -789.726     688.574
data_ref[T.2015-03-01]             224.3334    378.075      0.593      0.553    -516.753     965.420
data_ref[T.2015-04-01]             644.2383    376.941      1.709      0.087     -94.624    1383.101
data_ref[T.2015-05-01]             177.6984    375.907      0.473      0.636    -559.138     914.535
data_ref[T.2015-06-01]             642.5355    377.474      1.702      0.089     -97.373    1382.444
data_ref[T.2015-07-01]             368.1166    376.181      0.979      0.328    -369.256    1105.490
data_ref[T.2015-08-01]             210.0260    377.469      0.556      0.578    -529.872     949.924
data_ref[T.2015-09-01]             153.4899    374.937      0.409      0.682    -581.446     888.425
data_ref[T.2015-10-01]             270.5873    377.498      0.717      0.474    -469.367    1010.542
data_ref[T.2015-11-01]             445.9166    377.674      1.181      0.238    -294.384    1186.217
data_ref[T.2015-12-01]             621.5670    378.978      1.640      0.101    -121.289    1364.423
data_ref[T.2016-01-01]             302.8712    377.441      0.802      0.422    -436.972    1042.715
data_ref[T.2016-02-01]             -41.6680    377.355     -0.110      0.912    -781.342     698.006
data_ref[T.2016-03-01]              57.9217    380.707      0.152      0.879    -688.323     804.166
sexo[T.M]                         6054.7914    155.926     38.831      0.000    5749.152    6360.431
posse_de_veiculo[T.True]            48.1619    150.456      0.320      0.749    -246.755     343.079
posse_de_imovel[T.True]            363.7476    148.821      2.444      0.015      72.035     655.460
tipo_renda[T.Bolsista]           -1285.7818   2566.067     -0.501      0.616   -6315.672    3744.108
tipo_renda[T.Empresário]           829.2786    158.912      5.218      0.000     517.786    1140.771
tipo_renda[T.Pensionista]        -2725.7882   2566.941     -1.062      0.288   -7757.392    2305.816
tipo_renda[T.Servidor público]     138.7937    236.527      0.587      0.557    -324.836     602.424
educacao[T.Pós graduação]         1017.3601   1691.601      0.601      0.548   -2298.441    4333.161
educacao[T.Secundário]             570.3332    766.157      0.744      0.457    -931.455    2072.121
educacao[T.Superior completo]     1185.7821    769.133      1.542      0.123    -321.837    2693.402
educacao[T.Superior incompleto]    166.9395    827.796      0.202      0.840   -1455.669    1789.548
estado_civil[T.Separado]          1164.5763   1183.767      0.984      0.325   -1155.791    3484.944
estado_civil[T.Solteiro]           825.4338   1158.89

***2 - Remova a variável menos significante e analise***

***3 - Siga removendo as variáveis menos significantes, sempre que o p-value for menor que 5%. Compare o modelo final com o inicial. Observe os indicadores e conclua se o modelo parece melhor***

In [49]:
# criando matriz de design (e a matriz y) utilizando o Patsy, para um modelo em renda, explicado pelas demais variáveis, exclusive aquelas não significantes

y, X = patsy.dmatrices('renda ~ sexo + posse_de_imovel + idade + tempo_emprego', df)

reg_2 = sm.OLS(y,X).fit()

reg_2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  renda   R-squared:                       0.252
Model:                            OLS   Adj. R-squared:                  0.252
Method:                 Least Squares   F-statistic:                     1047.
Date:                Sat, 24 Jun 2023   Prob (F-statistic):               0.00
Time:                        21:33:52   Log-Likelihood:            -1.2882e+05
No. Observations:               12427   AIC:                         2.577e+05
Df Residuals:                   12422   BIC:                         2.577e+05
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
===========================================================================================
                              coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------------
Intercept               -2121.6370    329.140     -6.446      0.000   -2766.803   -1476.471
sexo[T.M]                6045.7894    145.591     41.526      0.000    5760.409    6331.169
posse_de_imovel[T.True]   394.7377    145.657      2.710      0.007     109.228     680.247
idade                      37.5533      7.838      4.791      0.000      22.189      52.918
tempo_emprego             550.4590     10.903     50.485      0.000     529.087     571.831
==============================================================================
Omnibus:                    17506.453   Durbin-Watson:                   2.036
Prob(Omnibus):                  0.000   Jarque-Bera (JB):          9670624.952
Skew:                           8.073   Prob(JB):                         0.00
Kurtosis:                     138.705   Cond. No.                         206.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

**Conclusão**: *O segundo modelo parece melhor. Além de ser muito menos complexo, apresenta um coefieciente de determinação R² (0.252) muito próximo ao apresentado pelo primeiro modelo (0.257)*